src/box/xrow.c

```diff
@@ -959,6 +959,15 @@ xrow_decode_dml(struct xrow_header *row, struct request *request,
 			request->after_tuple = value;
 			request->after_tuple_end = data;
 			break;
+		case IPROTO_SPACE_NAME:
+			key_map &= ~iproto_key_bit(IPROTO_SPACE_ID);
```

Comment:
This function is also used for decoding replication and recovery requests, where space names aren't accepted. I suggest to use the following approach insrtead:

<details><summary>diff</summary>

```diff
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index 22ed07ddd626..e93a876abf9b 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -1661,6 +1661,9 @@ iproto_msg_decode(struct iproto_msg *msg, struct cmsg_hop **route)
 		assert(type < sizeof(iproto_thread->dml_route) /
 			      sizeof(*iproto_thread->dml_route));
 		*route = iproto_thread->dml_route[type];
+		uint64_t key_map = dml_request_key_map(xrow->type);
+		/* space_id is optional - one may use space_name instead */
+		key_map |= 1ULL << IPROTO_SPACE_NAME;
 		if (xrow_decode_dml(&msg->header, &msg->dml,
 				    dml_request_key_map(type)))
 			return -1;
diff --git a/src/box/xrow.c b/src/box/xrow.c
index 0e71abb30468..ce72b0fbcb0d 100644
--- a/src/box/xrow.c
+++ b/src/box/xrow.c
@@ -960,7 +960,6 @@ error:
 			request->after_tuple_end = data;
 			break;
 		case IPROTO_SPACE_NAME:
-			key_map &= ~iproto_key_bit(IPROTO_SPACE_ID);
 			request->space_name =
 				mp_decode_str(&value, &request->space_name_len);
 			break;
@@ -973,6 +972,11 @@ error:
 		}
 	}
 done:
+	if ((key_map & (1ULL << IPROTO_SPACE_NAME)) != 0 &&
+	    request->space_name != NULL) {
+		/* The caller can accept space_name instead of space_id. */
+		key_map &= ~(1ULL << IPROTO_SPACE_ID);
+	}
 	if (key_map) {
 		enum iproto_key key = (enum iproto_key) bit_ctz_u64(key_map);
 		xrow_on_decode_err(row, ER_MISSING_REQUEST_FIELD,
```

</details>

Reply:
I didn't notice that we clear the bit for `IPROTO_SPACE_ID` above. Fixed the diff accordingly.

<details><summary>diff</summary>

```diff
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index a2ac55332a67..755f2c9b9166 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -1666,8 +1666,9 @@ iproto_msg_decode(struct iproto_msg *msg, struct cmsg_hop **route)
 		assert(type < sizeof(iproto_thread->dml_route) /
 			      sizeof(*iproto_thread->dml_route));
 		*route = iproto_thread->dml_route[type];
-		if (xrow_decode_dml(&msg->header, &msg->dml,
-				    dml_request_key_map(type)))
+		uint64_t key_map = dml_request_key_map(type) |
+				   iproto_key_bit(IPROTO_SPACE_NAME);
+		if (xrow_decode_dml(&msg->header, &msg->dml, key_map) != 0)
 			return -1;
 		/*
 		 * In contrast to replication requests, for a client request
diff --git a/src/box/xrow.c b/src/box/xrow.c
index 0e71abb30468..d1443254adad 100644
--- a/src/box/xrow.c
+++ b/src/box/xrow.c
@@ -890,7 +890,8 @@ error:
 		xrow_on_decode_err(row, ER_INVALID_MSGPACK, "packet body");
 		return -1;
 	}
-
+	bool space_name_is_accepted =
+		(key_map & iproto_key_bit(IPROTO_SPACE_NAME)) != 0;
 	uint32_t size = mp_decode_map(&data);
 	for (uint32_t i = 0; i < size; i++) {
 		if (mp_typeof(*data) != MP_UINT) {
@@ -973,7 +974,11 @@ error:
 		}
 	}
 done:
-	if (key_map) {
+	if (space_name_is_accepted && request->space_name != NULL) {
+		/* The caller can accept space_name instead of space_id. */
+		key_map &= ~iproto_key_bit(IPROTO_SPACE_ID);
+	}
+	if (key_map != 0) {
 		enum iproto_key key = (enum iproto_key) bit_ctz_u64(key_map);
 		xrow_on_decode_err(row, ER_MISSING_REQUEST_FIELD,
 				   iproto_key_name(key));
```

</details>

Reply:
Fixed. Added a test for recovery. @locker says we don't need a test for replication.